api/provider/aws/environment_update.go

```diff
@@ -33,6 +39,93 @@ func (e *EnvironmentProvider) Update(req models.UpdateEnvironmentRequest) error
 		}
 	}
 
+	if req.Links != nil {
+		sourceEnvSG, err := readSG(e.AWS.EC2, getEnvironmentSGName(fqEnvironmentID))
+		if err != nil {
+			return err
+		}
+
+		// check which environments are currently linked
+		currentTags, err := e.TagStore.SelectByTypeAndID("environment", environmentID)
+		if err != nil {
+			return err
+		}
+
+		currentLinkedEnvIDs := []string{}
+		for _, tag := range currentTags.WithKey("link") {
+			currentLinkedEnvIDs = strings.Split(tag.Value, ",")
+		}
+
+		checkForAWSError := func(err error) error {
+			if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "DoesNotExist" {
+				log.Printf("[WARN] skipping environment link/unlink since security group '%s' does not exist\n", getEnvironmentSGName(fqEnvironmentID))
+				return nil
+			}
+
+			return err
+		}
+
+		contains := func(value string, values []string) bool {
+			for _, v := range values {
+				if v == value {
+					return true
+				}
+			}
+
+			return false
+		}
+
+		// add - check if current link state contains links in request
+		for _, destEnvironmentID := range *req.Links {
+			if contains(destEnvironmentID, currentLinkedEnvIDs) {
+				continue
+			}
+
+			fqDestEnvID := addLayer0Prefix(e.Config.Instance(), destEnvironmentID)
+			sg, err := readSG(e.AWS.EC2, getEnvironmentSGName(fqDestEnvID))
+
+			if err != nil {
+				return checkForAWSError(err)
```

Comment:
this would return prematurely, when we actually don't want to return if `checkForAWSError()` return `nil` - we want to continue. Also, the check for idempontence should actually not occur here, since we do want to error when we can't add a security group that doesn't exist.

Reply:
I think we should actually return the error here without checking for the DoesNotExist as we're adding links in this loop. If the destination environment doesn't have an associated security group, we can't actually proceed so not sure what you mean when you say "we want to continue"?

I've updated it to
```
if err := e.createIngressInput(sourceSGID, destSGID); err != nil {
	return err
}
```